File: src/ts/src/io/xlsx.ts
/** XLSX (Microsoft Excel OOXML) 텍스트 추출 — Python ko_pii/io_/xlsx.py 대응.
 *
 * XLSX 구조:
 *   xl/sharedStrings.xml — 모든 inline strings (lookup)
 *   xl/worksheets/sheet1.xml ... — 셀 좌표 + 값/참조
 *
 * 각 ``<c>`` 셀의 ``t`` 속성:
 *   - ``s`` : sharedStrings 인덱스
 *   - ``str`` / ``inlineStr`` : 직접 문자열
 *   - 그 외 (숫자) : ``<v>`` 텍스트
 *
 * 테이블 분석을 위한 정형 추출 (`readRecords`) + 단순 텍스트 추출 (`readText`)
 * 둘 다 제공.
 */
import { pyIsAlpha } from "../core/strUtils.js";
import type { XmlElement } from "./xmlEt.js";
import { iterElements, parseXmlBytes, XmlParseError } from "./xmlEt.js";
import type { ZipFile } from "./zipFile.js";
import { openZip, zipNames, zipRead } from "./zipFile.js";

const MAIN_NS = "http://schemas.openxmlformats.org/spreadsheetml/2006/main";

/** Python ``root.findall("main:si", NS)`` — 직계 자식 중 main 네임스페이스 si. */
function findMainChildren(elem: XmlElement, local: string): XmlElement[] {
  return elem.children.filter((c) => c.nsUri === MAIN_NS && c.local === local);
}

/** Python ``c.find("main:v", NS)`` — 직계 자식 첫 번째. */
function findMainChild(elem: XmlElement, local: string): XmlElement | null {
  return elem.children.find((c) => c.nsUri === MAIN_NS && c.local === local) ?? null;
}

async function sharedStrings(zf: ZipFile): Promise<string[]> {
  if (!zipNames(zf).includes("xl/sharedStrings.xml")) return [];
  const root = parseXmlBytes(await zipRead(zf, "xl/sharedStrings.xml"));
  const strings: string[] = [];
  for (const si of findMainChildren(root, "si")) {
    // 각 <si> 안의 <t> 텍스트 (rich text 면 여러 개)
    const parts: string[] = [];
    for (const t of iterElements(si)) {
      if (t.nsUri === MAIN_NS && t.local === "t" && t.text) {
        parts.push(t.text);
      }
    }
    strings.push(parts.join(""));
  }
  return strings;
}

function cellValue(c: XmlElement, sst: string[]): string {
  const tAttr = c.attrs.t;
  if (tAttr === "s") {
    const v = findMainChild(c, "v");
    if (v === null || v.text === null) return "";
    // Python int(v.text) — 공백 strip + 부호 허용, 실패 시 ""
    const raw = v.text.trim();
    if (!/^[+-]?[0-9]+$/.test(raw)) return "";
    const idx = Number.parseInt(raw, 10);
    // Python list 인덱싱 — 음수는 뒤에서부터, 범위 밖은 ""
    const resolved = idx < 0 ? sst.length + idx : idx;
    return resolved >= 0 && resolved < sst.length ? (sst[resolved] ?? "") : "";
  }
  if (tAttr === "inlineStr") {
    const isElem = findMainChild(c, "is");
    if (isElem === null) return "";
    let out = "";
    for (const t of iterElements(isElem)) {
      if (t.nsUri === MAIN_NS && t.local === "t") {
        out += t.text ?? "";
      }
    }
    return out;
  }
  const v = findMainChild(c, "v");
  return v !== null && v.text !== null ? v.text : "";
}

/** 셀 참조('C5')의 0-based 열 인덱스. 참조 없으면 -1.
 *
 * 빈 셀은 XML 에서 ``<c>`` 가 생략되므로, 좌표를 무시하고 순서대로 이으면
 * 열이 왼쪽으로 밀려 정렬이 깨진다. r 속성으로 정확한 열에 배치한다.
 */
function colIndex(ref: string): number {
  let letters = "";
  for (const ch of ref) {
    if (pyIsAlpha(ch)) {
      letters += ch;
    } else if (letters) {
      break;
    }
  }
  if (!letters) return -1;
  let idx = 0;
  for (const ch of letters.toUpperCase()) {
    idx = idx * 26 + ((ch.codePointAt(0) ?? 0) - 64);
  }
  return idx - 1;
}

/** 탭/줄바꿈 구분된 텍스트로 모든 시트 + 셀 반환. */
export async function readText(path: string): Promise<string> {
  const parts: string[] = [];
  const zf = await openZip(path);
  const sst = await sharedStrings(zf);
  const sheetNames = zipNames(zf)
    .filter((n) => n.startsWith("xl/worksheets/sheet") && n.endsWith(".xml"))
    .sort();
  for (const name of sheetNames) {
    let root: XmlElement;
    try {
      root = parseXmlBytes(await zipRead(zf, name));
    } catch (e) {
      if (e instanceof XmlParseError) continue;
      throw e;
    }
    for (const row of iterElements(root)) {
      if (row.nsUri !== MAIN_NS || row.local !== "row") continue;
      const cells = new Map<number, string>();
      let seq = -1;
      for (const c of findMainChildren(row, "c")) {
        let ci = colIndex(c.attrs.r ?? "");
        if (ci < 0) ci = seq + 1;
        seq = Math.max(seq, ci);
        cells.set(ci, cellValue(c, sst));
      }
      const rowVals: string[] = [];
      for (let i = 0; i < seq + 1; i++) {
        rowVals.push(cells.get(i) ?? "");
      }
      parts.push(rowVals.join("\t"));
      parts.push("\n");
    }
    parts.push("\n"); // sheet separator
  }
  return parts.join("");
}

/** 첫 행을 헤더로, 이후 행을 ``{header: value}`` dict 리스트로 반환. */
export async function readRecords(path: string): Promise<Record<string, string>[]> {
  const text = await readText(path);
  const lines = text.split("\n").filter((ln) => ln.trim().length > 0);
  if (lines.length === 0) return [];
  const headers = lines[0]!.split("\t");
  const records: Record<string, string>[] = [];
  for (const ln of lines.slice(1)) {
    let cells = ln.split("\t");
    if (cells.length < headers.length) {
      cells = cells.concat(new Array(headers.length - cells.length).fill(""));
    }
    const record: Record<string, string> = {};
    // Python dict(zip(headers, cells[:len(headers)])) — 마지막 값이 이전 값을 덮음
    for (let i = 0; i < Math.min(headers.length, cells.length); i++) {
      record[headers[i]!] = cells[i]!;
    }
    records.push(record);
  }
  return records;
}

File: src/python/ko_pii/io_/xlsx.py
"""XLSX (Microsoft Excel OOXML) 텍스트 추출 — stdlib only.

XLSX 구조:
  xl/sharedStrings.xml — 모든 inline strings (lookup)
  xl/worksheets/sheet1.xml ... — 셀 좌표 + 값/참조

각 ``<c>`` 셀의 ``t`` 속성:
  - ``s`` : sharedStrings 인덱스
  - ``str`` / ``inlineStr`` : 직접 문자열
  - 그 외 (숫자) : ``<v>`` 텍스트

테이블 분석을 위한 정형 추출 (`read_records`) + 단순 텍스트 추출 (`read_text`)
둘 다 제공.
"""
from __future__ import annotations

import xml.etree.ElementTree as ET
import zipfile

NS = {"main": "http://schemas.openxmlformats.org/spreadsheetml/2006/main"}


def _shared_strings(zf: zipfile.ZipFile) -> list[str]:
    if "xl/sharedStrings.xml" not in zf.namelist():
        return []
    root = ET.fromstring(zf.read("xl/sharedStrings.xml"))
    strings: list[str] = []
    for si in root.findall("main:si", NS):
        # 각 <si> 안의 <t> 텍스트 (rich text 면 여러 개)
        parts = []
        for t in si.iter("{http://schemas.openxmlformats.org/spreadsheetml/2006/main}t"):
            if t.text:
                parts.append(t.text)
        strings.append("".join(parts))
    return strings


def _cell_value(c: ET.Element, sst: list[str]) -> str:
    t_attr = c.attrib.get("t")
    if t_attr == "s":
        v = c.find("main:v", NS)
        if v is None or v.text is None:
            return ""
        try:
            return sst[int(v.text)]
        except (ValueError, IndexError):
            return ""
    if t_attr == "inlineStr":
        is_elem = c.find("main:is", NS)
        if is_elem is None:
            return ""
        return "".join(t.text or "" for t in is_elem.iter(
            "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}t"))
    v = c.find("main:v", NS)
    return v.text if v is not None and v.text is not None else ""


def _col_index(ref: str) -> int:
    """셀 참조('C5')의 0-based 열 인덱스. 참조 없으면 -1.

    빈 셀은 XML 에서 ``<c>`` 가 생략되므로, 좌표를 무시하고 순서대로 이으면
    열이 왼쪽으로 밀려 정렬이 깨진다. r 속성으로 정확한 열에 배치한다.
    """
    letters = ""
    for ch in ref:
        if ch.isalpha():
            letters += ch
        elif letters:
            break
    if not letters:
        return -1
    idx = 0
    for ch in letters.upper():
        idx = idx * 26 + (ord(ch) - 64)
    return idx - 1


def read_text(path: str) -> str:
    """탭/줄바꿈 구분된 텍스트로 모든 시트 + 셀 반환."""
    parts: list[str] = []
    with zipfile.ZipFile(path, "r") as zf:
        sst = _shared_strings(zf)
        sheet_names = sorted(
            n for n in zf.namelist()
            if n.startswith("xl/worksheets/sheet") and n.endswith(".xml")
        )
        for name in sheet_names:
            try:
                root = ET.fromstring(zf.read(name))
            except ET.ParseError:
                continue
            for row in root.iter("{http://schemas.openxmlformats.org/spreadsheetml/2006/main}row"):
                cells: dict[int, str] = {}
                seq = -1
                for c in row.findall("main:c", NS):
                    ci = _col_index(c.attrib.get("r", ""))
                    if ci < 0:
                        ci = seq + 1
                    seq = max(seq, ci)
                    cells[ci] = _cell_value(c, sst)
                row_vals = [cells.get(i, "") for i in range(seq + 1)]
                parts.append("\t".join(row_vals))
                parts.append("\n")
            parts.append("\n")  # sheet separator
    return "".join(parts)


def read_records(path: str) -> list[dict[str, str]]:
    """첫 행을 헤더로, 이후 행을 ``{header: value}`` dict 리스트로 반환."""
    text = read_text(path)
    lines = [ln for ln in text.split("\n") if ln.strip()]
    if not lines:
        return []
    headers = lines[0].split("\t")
    records: list[dict[str, str]] = []
    for ln in lines[1:]:
        cells = ln.split("\t")
        if len(cells) < len(headers):
            cells = cells + [""] * (len(headers) - len(cells))
        records.append(dict(zip(headers, cells[:len(headers)])))
    return records
